File: liftoverKey.r
#!/usr/bin/Rscript

#Copyright (c) 2019 Brooke Wolford
# Labs of Dr. Cristen Willer and Dr. Mike Boehnke
# University of Michigan

options(stringsAsFactors=F)
library(data.table)
library(optparse)

option_list <- list(
    make_option("--old_bed", type="character", default=""),
    make_option("--new_bed",type="character",default=""),
    make_option("--unmapped",type="character",default=""),
    make_option("--prefix",type="character",default="")
)

parser <- OptionParser(usage="%prog [options]", option_list=option_list, description="This script creates a key for old reference genome coords to new reference genome coordinates after liftover\n")

args <- parse_args(parser, positional_arguments = 0)
opt <- args$options
print(opt)

old<-fread(opt$old_bed)
new<-fread(opt$new_bed)
unmapped<-data.table(read.table(opt$unmapped,comment.char="#"))

names(old)<-c("old_chr","old_poss","old_pose")
names(new)<-c("new_chr","new_poss","new_pose")
names(unmapped)<-c("un_chr","un_poss","un_pose")

old$old_coord<-paste(sep=":",old$old_chr,old$old_poss)
new$new_coord<-paste(sep=":",new$new_chr,new$new_poss)
unmapped$coord<-paste(sep=":",unmapped$un_chr,unmapped$un_poss)

#initialize variables as we are adding columns to the old file
old$new_chr<-as.character(NA)
old$new_poss<-as.numeric(NA)
old$new_pose<-as.numeric(NA)
old$new_coord<-as.character(NA)

for (i in 1:nrow(old)){
    if (old[i]$old_coord %in% unmapped$coord){
        next
    } else{
        old[i]$new_chr<-new[i]$new_chr
        old[i]$new_poss<-new[i]$new_poss
        old[i]$new_pose<-new[i]$new_pose
        old[i]$new_coord<-new[i]$new_coord
    }
    
}
 
#write output
filename<-paste(sep=".",opt$prefix,"txt")
write.table(old, file=filename,sep="\t",append=FALSE,quote=FALSE)

File: liftoverKey.py
#!/usr/bin/python3

#Script by Brooke Wolford
## December 13, 2019

import gzip
import sys
import re
import argparse
import linecache
from collections import OrderedDict

### Written for Python 3

argparser = argparse.ArgumentParser('This script creates a key for old reference genome coords to new reference genome coordinates after liftover\n')
argparser.add_argument("--old_bed",type=str,help="Old coordinate bed file that was submitted to liftOver")
argparser.add_argument("--new_bed",type=str,help="New coordinate bed file from liftOver")
argparser.add_argument("--unmapped",type=str,help="Unmapped file from liftOver")
argparser.add_argument("--prefix",type=str,help="Prefix to add to .txt for output")

def open_zip(f):
    """open files"""
    if ".gz" in f:
        command=gzip.open(f,"rt")
        print("Opening gzipped file %s\n" % f,file=sys.stderr)
    elif f == "-":
        sys.exit("Cannot read file from stdin\n")
    else:
        command=open(f,"rt")
        print("Opening file %s\n" % f, file=sys.stderr)
        return command
    
def read_bed(f):
    command=open_zip(f)
    my_dict=OrderedDict()
    with command as bed_file:
        for line in bed_file:
            ls=line.rstrip()
            if not ls.startswith("#"): #ignore comment lines 
                line_list=ls.split("\t")
                my_dict[":".join([line_list[0],line_list[1],line_list[2]])]=line_list
    bed_file.close()
    return(my_dict)


def make_key(new,old,unmap,prefix):
    command=open_zip(old)
    ucount=0 #unampped count
    lcount=0 #line count
    results_filename=".".join([prefix,"txt"])
    with command as new_file:
        for line in new_file:
            ls=line.rstrip()
            line_list=ls.split("\t")
            old_coord=":".join([line_list[0],line_list[1],line_list[2]])
            with open(results_filename, "w") as out:
                out.write("\t".join("old_chrom","old_posStart","old_posEnd","old_coord","new_chrom","new_posStart","new_posEnd","new_coord"))
                if old_coord in unmap: #old coordiante didn't get mapped so NA for key
                    ucount+=1
                    lcount+=1
                    out.write("\t".join(["\t".join(line_list),old_coord,"\t".join(["NA","NA","NA","NA"])]))
                else:
                    lcount+=1
                    nl=linecache.getline(new,lcount-ucount) #new line
                    nl_list=nl.rstrip().split("\t")
                    out.write("\t".join(["\t".join(line_list),old_coord,"\t".join(nl_list),":".join([nl_list[0],nl_list[1],nl_list[2]])]))

    new_file.close()
    out.close()

###############################
######## MAIN ###############
############################
            
if __name__ == '__main__':
    args = argparser.parse_args()
    print(args, file=sys.stderr)
    
    unmapped=read_bed(args.unmapped)
    make_key(args.new_bed,args.old_bed,unmapped,args.prefix)

